fem/estimators.cpp

```diff
@@ -207,115 +211,88 @@ void KellyErrorEstimator::ComputeEstimates()
 
       ElementTransformation* Transf = xfes->GetElementTransformation(e);
       flux_integrator->ComputeElementFlux(*xfes->GetFE(e), *Transf, el_x,
-                                          *flux_space->GetFE(e), el_f, true);
+                                          *flux_space->GetFE(e), el_f, with_coef);
 
       flux_space->GetElementVDofs(e, fdofs);
       flux->AddElementVector(fdofs, el_f);
    }
 
    // 2. Add error contribution from local interior faces
-   for (int f = 0; f < mesh->GetNumFaces(); f++)
+   IntegrationPoint ip;
+   Vector val(flux_space->GetVDim());
+   Vector normal(mesh->SpaceDimension());
+   Vector ref_normal(mesh->Dimension());
+   for (int fi = 0; fi < mesh->GetNumFaces(); fi++)
    {
-      auto FT = mesh->GetFaceElementTransformations(f);
-
-      auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(f));
-      const auto nip = int_rule.GetNPoints();
-
-      if (mesh->FaceIsInterior(f))
+      if (mesh->FaceIsInterior(fi))
       {
-         int Inf1, Inf2, NCFace;
-         mesh->GetFaceInfos(f, &Inf1, &Inf2, &NCFace);
-
-         // Convention
-         // * Conforming face: Face side with smaller element id handles
-         // the integration
-         // * Non-conforming face: The slave handles the integration.
-         // See FaceInfo documentation for details.
-         bool isNCSlave    = FT->Elem2No >= 0 && NCFace >= 0;
-         bool isConforming = FT->Elem2No >= 0 && NCFace == -1;
-         if ((FT->Elem1No < FT->Elem2No && isConforming) || isNCSlave)
+         // Compute NC and face information
+         int FaceElement1, FaceElement2, NCFace;
+         mesh->GetFaceInfos(fi, &FaceElement1, &FaceElement2, &NCFace);
+         mesh->GetFaceElements(fi, &FaceElement1, &FaceElement2);
+
+         // We skip over master faces
+         bool isNCSlave    = FaceElement2 >= 0 && NCFace >= 0;
+         bool isConforming = FaceElement2 >= 0 && NCFace == -1;
+         if (isConforming || isNCSlave)
          {
             if (attributes.Size() &&
-                (attributes.FindSorted(FT->Elem1->Attribute) == -1
-                 || attributes.FindSorted(FT->Elem2->Attribute) == -1))
+                (attributes.FindSorted(mesh->GetAttribute(FaceElement1)) == -1
+                 || attributes.FindSorted(mesh->GetAttribute(FaceElement2)) == -1))
             {
                continue;
             }
 
-            IntegrationRule eir;
-            Vector jumps(nip);
+            // Computation of the transformations is not super cheap, so we delay it until here. All ops above are cheaper lookups.
+            auto FT = mesh->GetFaceElementTransformations(fi);
+
+            // TODO Refactor this computation into a user-facing function.
+            // auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(fi)); // NOTE: This fails for DG
```

Comment:
Thanks!

I am currently preparing a publication where I use Kelly for AMR with DG discretization and it works really better than expected for my problem, although I cannot give a general recommendation to always use Kelly. The problem here is that MFEM segfaults in the call to `xfes->GetFaceOrder(fi)` if xfes is $L_2$. From my understanding the segfault is triggered due to $L_2$ spaces not having an interpolation on faces defined. I think the current DG diffusion integrator has the same issue as this code - see https://github.com/mfem/mfem/blob/f97a770dcaa7498c93069ca7f27d25d9655acd86/fem/bilininteg.cpp#L3424-L3434 .

If you have a better solution in mind I am happy to implement it!

Reply:
Thanks for taking time for the valuable comment @hughcars . Correct me if I am wrong, but the face order would be the interpolation order constrained to the face, right? Here, for mixed order H1 spaces it would be the minimal order of the space, because of the continuity constraint, however for mixed order L2 spaces it whould be the maximum order, or am I missing something?

Reply:
Yes sorry I got it the wrong way round, in H1 there would be the NC constraints activated, so the effective order of the higher space would be reduced, whereas for L2 there would not be, so the selection would be the higher order.